drivers/note/note_lastsched_driver.c

```diff
@@ -0,0 +1,397 @@
+                                    FAR struct tcb_s *tcb, bool enter)
+{
+  note_lastsched_common(
+      tcb, enter ? NOTE_CSECTION_ENTER : NOTE_CSECTION_LEAVE, NULL);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SPINLOCKS
+static void note_spinlock(FAR struct note_driver_s *drv,
+                          FAR struct tcb_s *tcb, FAR volatile void *spinlock,
+                          int type)
+{
+  note_lastsched_common(tcb, type, (FAR void *)spinlock);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+static void note_lastsched_irqhandler(FAR struct note_driver_s *drv, int irq,
+                                      FAR void *handler, bool enter)
+{
+  note_lastsched_common(this_task(), enter ? NOTE_IRQ_ENTER : NOTE_IRQ_LEAVE,
+                        (FAR void *)irq);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+static void note_lastsched_syscall_enter(FAR struct note_driver_s *drv,
+                                         int nr, int argc, va_list *ap)
+{
+  note_lastsched_common(this_task(), NOTE_SYSCALL_ENTER, (void *)nr);
+}
+
+static void note_lastsched_syscall_leave(FAR struct note_driver_s *drv,
+                                         int nr, uintptr_t result)
+{
+  note_lastsched_common(this_task(), NOTE_SYSCALL_LEAVE, (void *)nr);
+}
+#endif
+
+/****************************************************************************
+ * Name: note_lastched_register
+ ****************************************************************************/
+
+int note_lastched_register(void)
+{
+  g_note_lastsched.freq = up_perf_getfreq();
+  if (g_note_lastsched.freq == 0)
+    {
+      g_note_lastsched.freq = NSEC_PER_SEC;
+    }
+
+  clock_gettime(CLOCK_REALTIME, &g_note_lastsched.ts);
+  return note_driver_register(&g_note_lastsched_driver);
+}
+
+/****************************************************************************
+ * Name: note_lastsched_dump
+ ****************************************************************************/
+
+void note_lastsched_dump(void)
+{
+  size_t i;
+  uint32_t freq;
+  irqstate_t flags;
+  struct note_filter_mode_s old;
+  size_t index = g_note_lastsched.index;
+  FAR struct note_lastsched_chunk_s *note;
+
+  sched_note_filter_mode(&old, NULL);
+  old.flag &= ~(NOTE_FILTER_MODE_FLAG_ENABLE);
+  sched_note_filter_mode(NULL, &old);
+  flags = enter_critical_section();
+
+  syslog(LOG_INFO,
+         "last_sched: system frequency: %" PRIu32 "\n"
+         "last_sched: start time: %" PRIu32 ".%" PRIu32,
+         g_note_lastsched.freq,
+         g_note_lastsched.ts.tv_sec,
+         g_note_lastsched.ts.tv_nsec);
+
+  freq = g_note_lastsched.freq ? NSEC_PER_SEC : g_note_lastsched.freq;
+  for (i = index; i != index - 1; i++)
+    {
+      note = &g_note_lastsched.buffer[i];
+      i = (i == CONFIG_DRIVER_NOTE_LASTSCHED_NBUFFERS - 1) ? 0 : i;
+      syslog(LOG_INFO,
+             "last_sched|%" PRIuPTR ": time:%" PRIu32 ", cpu: %" PRIu32 ", "
```

Comment:
let's follow syslog output compact format.

Reply:
done

Reply:
It seems more reasonable to change notelast to notesched
All it records is scheduling information, which can be used not only in panic, but also added to the trace command

Reply:
Do you plan to expose this driver to user space? which will dup many code in noteram driver, and I don't see any benefit to not use ramnote driver for this.